### PR TITLE
run at most n things concurrently

### PR DESCRIPTION
**Pull request by toolslive**

### CODE REVIEW

File: ocaml/src/maintenance.ml [hunk 1]
```diff
@@ -741,25 +753,29 @@ class client ?(retry_timeout = 60.)
...
-                    >>= fun (_, devices, _) ->
-                    List.iter
-                      (fun osd_id ->
-                       if not (Hashtbl.mem threads osd_id)
-                       then begin
-                           Hashtbl.add threads osd_id ();
-                           Lwt.async
-                             (fun () ->
-                              Lwt_extra2.ignore_errors
+                    Lwt_pool2.use
```
Comment (domsj) [inline ocaml/src/maintenance.ml]: Push this throttling down to (or around) the `self # garbage_collect_device` call?

**Comment by domsj** [comment]
`decommission_device` and some items in `handle_work_item` (specifically RepairBadFragment & IterNamespaceLeaf, not the others?) can use some throttling too...

**Comment by domsj** [comment]
I would wrap all the calls to `Repair.rewrite_object` & `Verify.verify_and_maybe_repair_object` (to request a 'resource' first).
The question at https://github.com/openvstorage/alba/pull/102#discussion_r47234341 remains.

**Comment by toolslive** [comment]
I don't think `garbage_collect_device` needs to be throttled more than it already does by itself.

**Comment by domsj** [comment]
You're right that garbage_collect_device doesn't need any throttling, so I suggest removing it. For the rest: looks good to me.

**Pull request merged**